sdks/python/apache_beam/testing/analyzers/perf_analysis_utils.py

```diff
@@ -0,0 +1,199 @@
+  else return latest created issue_number along with
+  """
+  query = f"""
+  SELECT * FROM {constants._BQ_PROJECT_NAME}.{constants._BQ_DATASET}.{test_name}
+  ORDER BY {constants._ISSUE_CREATION_TIMESTAMP_LABEL} DESC
+  LIMIT 10
+  """
+  try:
+    df = big_query_metrics_fetcher.fetch(query=query)
+  except exceptions.NotFound:
+    # If no table found, that means this is first performance regression
+    # on the current test+metric.
+    return None
+  return df
+
+
+def is_perf_alert(
+    previous_change_point_timestamps: List[pd.Timestamp],
+    change_point_index: int,
+    timestamps: List[pd.Timestamp],
+    min_runs_between_change_points: int) -> bool:
+  """
+  Search the previous_change_point_timestamps with current observed
+  change point sibling window and determine if it is a duplicate
+  change point or not.
+
+  Return False if the current observed change point is a duplicate of
+  already reported change points else return True.
+  """
+  sibling_change_point_min_timestamp = timestamps[min(
+      change_point_index + min_runs_between_change_points, len(timestamps) - 1)]
+  sibling_change_point_max_timestamp = timestamps[max(
+      0, change_point_index - min_runs_between_change_points)]
+  # Search a list of previous change point timestamps and compare it with
+  # current change point timestamp. We do this in case, if a current change
+  # point is already reported in the past.
+  for previous_change_point_timestamp in previous_change_point_timestamps:
+    if (sibling_change_point_min_timestamp <= previous_change_point_timestamp <=
+        sibling_change_point_max_timestamp):
+      return False
+  return True
+
+
+def read_test_config(config_file_path: str) -> Dict:
+  """
+  Reads the config file in which the data required to
+  run the change point analysis is specified.
+  """
+  with open(config_file_path, 'r') as stream:
+    config = yaml.safe_load(stream)
+  return config
+
+
+def validate_config(keys):
+  return constants._PERF_TEST_KEYS.issubset(keys)
+
+
+def fetch_metric_data(
+    params: Dict[str, Any], big_query_metrics_fetcher: BigQueryMetricsFetcher
+) -> Tuple[List[Union[int, float]], List[pd.Timestamp]]:
+  query = f"""
+      SELECT *
+      FROM {params['project']}.{params['metrics_dataset']}.{params['metrics_table']}
+      WHERE CONTAINS_SUBSTR(({load_test_metrics_utils.METRICS_TYPE_LABEL}), '{params['metric_name']}')
+      ORDER BY {load_test_metrics_utils.SUBMIT_TIMESTAMP_LABEL} DESC
+      LIMIT {constants._NUM_DATA_POINTS_TO_RUN_CHANGE_POINT_ANALYSIS}
+    """
+  metric_data: pd.DataFrame = big_query_metrics_fetcher.fetch(query=query)
+  return (
+      metric_data[load_test_metrics_utils.VALUE_LABEL],
+      metric_data[load_test_metrics_utils.SUBMIT_TIMESTAMP_LABEL])
+
+
+def find_latest_change_point_index(metric_values: List[Union[float, int]]):
+  change_points_idx = e_divisive(metric_values)
+  if not change_points_idx:
+    return None
+  # Consider the latest change point.
```

Comment:
I'm having a hard time reasoning about the order of elements here.

fetch_metric_data returns elements in descending order of the timestamp (new timestamps first), same order applies to metric_values.

find_latest_change_point_index  gets indices returned by e_devisive, sorts them in descending ascending order (larger indicies first), then returns, first element. Larger indices would correspond to older timestamps. End result seems like the oldest changepoint to me, not the latest (more recent changepoint)

Note that order of elements, again, is not documented.

Reply:
You are right. This confuses a users. I changed metric_values and timestamps to be sorted based on timestamps.

Reply:
I am still seeing 
`      ORDER BY {load_test_metrics_utils.SUBMIT_TIMESTAMP_LABEL} DESC` though in the query?